File: sdk/ruby/lib/phantex/events.rb
# frozen_string_literal: true

require "digest"
require "json"
require "securerandom"

module Phantex
  # Event type codes — matches proto/phantex/v1/events.proto.
  module EventType
    UNSPECIFIED        = 0
    PROCESS_EXEC       = 1
    PROCESS_EXIT       = 2
    FILE_OPEN          = 10
    FILE_WRITE         = 11
    FILE_READ          = 12
    NETWORK_CONNECT    = 20
    NETWORK_ACCEPT     = 21
    NETWORK_DNS        = 22
    MEMORY_MMAP        = 30
    AGENT_DISCOVERED   = 40
    AGENT_TERMINATED   = 41
    TOOL_CALL          = 50
    TOOL_RESPONSE      = 51
    ALERT_FIRED        = 60
  end

  # Severity levels — matches proto/phantex/v1/events.proto.
  module Severity
    UNSPECIFIED = 0
    INFO        = 1
    LOW         = 2
    MEDIUM      = 3
    HIGH        = 4
    CRITICAL    = 5
  end

  # Base event with common envelope fields.
  class Event
    attr_accessor :event_id, :tenant_id, :agent_id, :sensor_id,
                  :event_type, :severity, :timestamp_ns,
                  :trace_id, :span_id, :parent_span_id,
                  :framework, :extra

    def initialize(event_type:, **kwargs)
      @event_id       = kwargs[:event_id]       || SecureRandom.hex(16)
      @tenant_id      = kwargs[:tenant_id]      || ""
      @agent_id       = kwargs[:agent_id]       || ""
      @sensor_id      = kwargs[:sensor_id]      || ""
      @event_type     = event_type
      @severity       = kwargs[:severity]        || Severity::INFO
      @timestamp_ns   = kwargs[:timestamp_ns]    || (Time.now.to_f * 1_000_000_000).to_i
      @trace_id       = kwargs[:trace_id]        || ""
      @span_id        = kwargs[:span_id]         || ""
      @parent_span_id = kwargs[:parent_span_id]  || ""
      @framework      = kwargs[:framework]       || ""
      @extra          = kwargs[:extra]           || {}
    end

    def to_h
      {
        event_id:       @event_id,
        tenant_id:      @tenant_id,
        agent_id:       @agent_id,
        sensor_id:      @sensor_id,
        event_type:     @event_type,
        severity:       @severity,
        timestamp_ns:   @timestamp_ns,
        trace_id:       @trace_id,
        span_id:        @span_id,
        parent_span_id: @parent_span_id,
        framework:      @framework,
      }.merge(@extra)
    end

    def to_json(*_args)
      JSON.generate(to_h)
    end
  end

  # Tool call event — emitted when an AI framework invokes a tool.
  class ToolCallEvent < Event
    attr_accessor :tool_name, :tool_input, :protocol

    def initialize(tool_name:, protocol: "unknown", tool_input: nil, **kwargs)
      super(event_type: EventType::TOOL_CALL, **kwargs)
      @tool_name  = tool_name
      @protocol   = protocol
      @tool_input = tool_input
    end

    def to_h
      super.merge(
        tool_name:  @tool_name,
        protocol:   @protocol,
        tool_input: safe_serialize(@tool_input),
      )
    end

    private

    def safe_serialize(obj, max_bytes: 4096)
      return "" if obj.nil?

      raw = JSON.generate(obj)
      return raw if raw.bytesize <= max_bytes

      # Truncate by bytes, then ensure we don't split a multi-byte character
      truncated = raw.byteslice(0, max_bytes - 3)&.scrub("") || ""
      "#{truncated}..."
    rescue StandardError
      "<unserializable>"
    end
  end

  # Tool response event — emitted after a tool call completes.
  class ToolResponseEvent < Event
    attr_accessor :tool_name, :protocol, :success, :duration_ns, :error_message

    def initialize(tool_name:, protocol: "unknown", success: true, duration_ns: 0, error_message: nil, **kwargs)
      super(event_type: EventType::TOOL_RESPONSE, **kwargs)
      @tool_name     = tool_name
      @protocol      = protocol
      @success       = success
      @duration_ns   = duration_ns
      @error_message = error_message
    end

    def to_h
      h = super.merge(
        tool_name:   @tool_name,
        protocol:    @protocol,
        success:     @success,
        duration_ns: @duration_ns,
      )
      h[:error_message] = @error_message if @error_message
      h
    end
  end

  # Hash prompt content — never store plaintext.
  def self.hash_prompt(prompt)
    Digest::SHA256.hexdigest(prompt.to_s)
  end
end

File: sdk/python/phantex_sdk/events.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2025-2026 The Phantex Authors

"""
Phantex SDK event definitions.

Matches the protobuf schema (events.proto) for tool call / tool response events.
Events are plain dataclasses — no protobuf dependency required at runtime.
Serialized to JSON dicts for transport.
"""

from __future__ import annotations

import hashlib
import json
import time
import uuid
from dataclasses import asdict, dataclass, field
from enum import IntEnum
from typing import Any

# ── Event Types (mirrors EventType in events.proto) ──────────────────────────

class EventType(IntEnum):
    """Event type codes — must match proto/phantex/v1/events.proto."""

    UNSPECIFIED = 0
    PROCESS_EXEC = 1
    PROCESS_EXIT = 2
    FILE_OPEN = 10
    FILE_WRITE = 11
    FILE_READ = 12
    NETWORK_CONNECT = 20
    NETWORK_ACCEPT = 21
    NETWORK_DNS = 22
    MEMORY_MMAP = 30
    AGENT_DISCOVERED = 40
    AGENT_TERMINATED = 41
    TOOL_CALL = 50
    TOOL_RESPONSE = 51
    ALERT_FIRED = 60

class Severity(IntEnum):
    """Severity levels — must match proto/phantex/v1/events.proto."""

    UNSPECIFIED = 0
    INFO = 1
    LOW = 2
    MEDIUM = 3
    HIGH = 4
    CRITICAL = 5

# ── Helper Functions ──────────────────────────────────────────────────────────

def _event_id() -> str:
    """Generate a UUID v4 event ID (hex, no dashes)."""
    return uuid.uuid4().hex

def _timestamp_ns() -> int:
    """Current time in nanoseconds since epoch."""
    return int(time.time() * 1_000_000_000)

def _hash_prompt(prompt: str) -> str:
    """SHA-256 hash of prompt content. Never store plaintext prompts."""
    return hashlib.sha256(prompt.encode("utf-8", errors="replace")).hexdigest()

def _safe_serialize(obj: Any, max_bytes: int = 4096) -> str:
    """
    Safely serialize an object to JSON string, truncated to max_bytes.
    Never raises — returns "<unserializable>" on failure.
    """
    try:
        raw = json.dumps(obj, default=str, ensure_ascii=False)
        if len(raw) > max_bytes:
            return raw[: max_bytes - 3] + "..."
        return raw
    except Exception:
        return "<unserializable>"

# ── Event Dataclasses ─────────────────────────────────────────────────────────

@dataclass(slots=True)
class ToolCallEvent:
    """
    A tool/function call captured by an SDK hook.

    Maps to EventType.TOOL_CALL (50) in the protobuf schema.
    """

    # ── Identity ──────────────────────────────────
    event_id: str = field(default_factory=_event_id)
    event_type: int = field(default=EventType.TOOL_CALL, init=False)
    timestamp_ns: int = field(default_factory=_timestamp_ns)

    # ── Agent/Tenant context ──────────────────────
    tenant_id: str = ""
    agent_paid: str = ""
    pid: int = 0

    # ── Tool info ─────────────────────────────────
    tool_name: str = ""
    tool_input: str = ""  # Serialized arguments (truncated, redacted)
    protocol: str = ""  # "langchain_tool", "autogen", "crewai", "http", etc.
    framework: str = ""  # "langchain", "autogen", "crewai"

    # ── LLM call info (for HTTP / ChatModel hooks) ─
    model_name: str = ""  # e.g., "gpt-4", "claude-3-opus"
    prompt_hash: str = ""  # SHA-256 of prompt content
    input_tokens: int = 0
    output_tokens: int = 0

    # ── Trace context ─────────────────────────────
    trace_id: str = ""
    span_id: str = ""
    parent_span_id: str = ""

    # ── Severity ──────────────────────────────────
    severity: int = field(default=Severity.INFO)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dict for JSON transport. Keeps False/0 when semantically meaningful."""
        d = {}
        for k, v in asdict(self).items():
            # Keep False booleans and 0 integers (semantically meaningful)
            # Drop empty strings, None, and empty collections
            if v is not None and v != "" and v != []:
                d[k] = v
        return d

@dataclass(slots=True)
class ToolResponseEvent:
    """
    A tool/function response captured by an SDK hook.

    Maps to EventType.TOOL_RESPONSE (51) in the protobuf schema.
    """

    # ── Identity ──────────────────────────────────
    event_id: str = field(default_factory=_event_id)
    event_type: int = field(default=EventType.TOOL_RESPONSE, init=False)
    timestamp_ns: int = field(default_factory=_timestamp_ns)

    # ── Agent/Tenant context ──────────────────────
    tenant_id: str = ""
    agent_paid: str = ""
    pid: int = 0

    # ── Tool info ─────────────────────────────────
    tool_name: str = ""
    protocol: str = ""
    framework: str = ""

    # ── Result ────────────────────────────────────
    success: bool = True
    duration_ns: int = 0  # Wall-clock time in nanoseconds
    output_size: int = 0  # Response size in bytes (content NOT captured)
    error_message: str = ""  # Only set on failure

    # ── LLM response info ─────────────────────────
    model_name: str = ""
    input_tokens: int = 0
    output_tokens: int = 0

    # ── Trace context ─────────────────────────────
    trace_id: str = ""
    span_id: str = ""
    parent_span_id: str = ""

    # ── Severity ──────────────────────────────────
    severity: int = field(default=Severity.INFO)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dict for JSON transport. Keeps False/0 when semantically meaningful."""
        d = {}
        for k, v in asdict(self).items():
            if v is not None and v != "" and v != []:
                d[k] = v
        return d
